src/compiler/checker.ts

```diff
@@ -38263,13 +38263,14 @@ namespace ts {
         function isReferencedAliasDeclaration(node: Node, checkChildren?: boolean): boolean {
             if (isAliasSymbolDeclaration(node)) {
                 const symbol = getSymbolOfNode(node);
-                if (symbol && getSymbolLinks(symbol).referenced) {
+                const links = symbol && getSymbolLinks(symbol);
+                if (links?.referenced || (compilerOptions.isolatedModules && links?.constEnumReferenced)) {

```

Comment:
I am not sure why this change is needed ? Replacing `shouldPreserveConstEnums` instead of `options.preserveConstEnums` should mark things correctly if not why is it not a issue with preserveConstEnums? Does necessity of this change means somewhere we are not using `shouldPreserveConstEnums`

Reply:
I missed this—I think it may be left over from before and could be changed now. @FauxFaux?

Reply:
This is still the actual bug fix, the only code that needed to be changed.

`--preserveConstEnums` stops the compiler deleting the enum definition (but still inlines uses)
`--isolatedModules` prevents the inlining of the const enum, so requires the enum definition to be not deleted, *and* the reference imported?

This line implements the "and imported" fix.

On the released compiler:

```
% tsc --version
Version 4.1.3
```

### inlined
```
% tsc --target es2018 file1.ts; head -n 10 file?.js
==> file1.js <==
console.log(0 /* BAR */);
export {};

==> file2.js <==
export {};
```

### still inlined
```
% tsc --preserveConstEnums file1.ts; head -n 10 file?.js
==> file1.js <==
"use strict";
exports.__esModule = true;
console.log(0 /* BAR */);

==> file2.js <==
"use strict";
exports.__esModule = true;
exports.Foo = void 0;
var Foo;
(function (Foo) {
    Foo[Foo["BAR"] = 0] = "BAR";
})(Foo = exports.Foo || (exports.Foo = {}));
```

### not inlined but not imported
```
% tsc --isolatedModules file1.ts; head -n 10 file?.js   
==> file1.js <==
"use strict";
exports.__esModule = true;
console.log(file2_1.Foo.BAR);

==> file2.js <==
"use strict";
exports.__esModule = true;
exports.Foo = void 0;
var Foo;
(function (Foo) {
    Foo[Foo["BAR"] = 0] = "BAR";
})(Foo = exports.Foo || (exports.Foo = {}));
```

This PR behaves exactly like the released compiler if you take this hunk out.

(The relevant decision appears to be inside `markAliasReferenced` on a five digit line number in `checker.ts`, a file so big I apparently can't link to it, and can barely edit.)

Reply:
`isolatedModules` should behave exactly like `preserveConstEnums`

Reply:
IIUC `isolatedModules` should behave like a _superset_ of `preserveConstEnums`—`preserveConstEnums` doesn’t prevent the inlining of const enum values, it only prevents the erasure of the declaration. `isolatedModules` must do both. Or in other words, `isolatedModules` is the _only_ option that prevents the inlining of const enums. Therefore, for a const enum alias to be considered referenced, it _must_ be referenced under `isolatedModules`. This line is correct—sorry it took me a minute to remember the nuances after the holidays 😅

Reply:
Yes, that's what I was trying to say! :+1:

Reply:
If thats the case this is not the correct place to fix it ? shouldnt it be handled in `checkPropertyAccessExpressionOrQualifiedName` when isolatedModules is set

Reply:
`constEnumReferenced` is for typeonly imports and shouldnt be mixed

Reply:
Hmm, I forgot I did add that property for the purpose of type-only imports, but I think it’s a reasonably good fit for this. But it may make just as much sense to mark const enum aliases as `referenced`, not just `constEnumReferenced`, under `isolatedModules` in `markAliasReferenced`.

That said, the condition in `checkPropertyAccessExpressionOrQualifiedName` probably does cause a bug in this implementation, though, so that will need to be addressed one way or another. `tests/cases/compiler/constEnumNamespaceReferenceCausesNoImport.ts` should be duplicated and tested under `isolatedModules` as well.

Reply:
(Instead of duplicating, you can add `// @isolatedModules: true,false` to the top)

Reply:
That test is indeed bad under isolatedModules:true.

I do not understand any of the other methods you referenced, so I am not intending to make any further changes to this PR.